### ISSUE TITLE
It should be possible to change the cardinality of a feature in a domain without losing existing data

### ISSUE THREAD

**Issue by pcdavid** [eclipse-sirius/sirius-components#615]
Sirius EMF JSON is as robust to small "metamodel" changes as it could be.

### Steps to reproduce

1. Create a new project `studio`. The studio definition feature must of course be enabled.
2. Create a new domain with two entities `Root` and `Child`
3. Add a containment relation `Root.child : Child`. Keep it single-valued.
4. Create a new project, `instance`.
5. In it, create an instance of `Root` and a `Child` inside of it.
6. Close the `instance` project. You might need to wait a little bit (unless you customized `org.eclipse.sirius.web.editingContextEventProcessorRegistry.disposeDelay` to a low enough value).
7. In the `studio` project, make the `child` relation multi-valued.
8. Re-open the `instance`: it would be natural to still have the `Child` element inside `Root`, but it is gone.

The backend's log shows this error:
```
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Not a JSON Array: {"id":"cbcb071a-0d7b-47b2-9486-7040c0cdf354","eClass":"sampledomain:Child"}
	at com.google.gson.Gson.fromJson(Gson.java:944)
	at org.eclipse.sirius.emfjson.resource.JsonResourceImpl.doLoad(JsonResourceImpl.java:299)
	at org.eclipse.emf.ecore.resource.impl.ResourceImpl.load(ResourceImpl.java:1563)
	at org.eclipse.sirius.web.services.editingcontext.EditingContextSearchService.findById(EditingContextSearchService.java:109)
	at org.eclipse.sirius.web.spring.collaborative.projects.EditingContextEventProcessorRegistry.getOrCreateEditingContextEventProcessor(EditingContextEventProcessorRegistry.java:87)
	at org.eclipse.sirius.web.graphql.datafetchers.subscriptions.SubscriptionTreeEventDataFetcher.get(SubscriptionTreeEventDataFetcher.java:80)
	at org.eclipse.sirius.web.graphql.datafetchers.subscriptions.SubscriptionTreeEventDataFetcher.get(SubscriptionTreeEventDataFetcher.java:1)
	at graphql.execution.instrumentation.dataloader.DataLoaderDispatcherInstrumentation.lambda$instrumentDataFetcher$0(DataLoaderDispatcherInstrumentation.java:86)
	at graphql.execution.ExecutionStrategy.fetchField(ExecutionStrategy.java:270)
	at graphql.execution.SubscriptionExecutionStrategy.createSourceEventStream(SubscriptionExecutionStrategy.java:91)
	at graphql.execution.SubscriptionExecutionStrategy.execute(SubscriptionExecutionStrategy.java:53)
	at graphql.execution.Execution.executeOperation(Execution.java:165)
	at graphql.execution.Execution.execute(Execution.java:104)
	at graphql.GraphQL.execute(GraphQL.java:557)
	at graphql.GraphQL.parseValidateAndExecute(GraphQL.java:482)
	at graphql.GraphQL.executeAsync(GraphQL.java:446)
	at graphql.GraphQL.execute(GraphQL.java:377)
	at org.eclipse.sirius.web.spring.graphql.ws.handlers.StartMessageHandler.handle(StartMessageHandler.java:101)
	at org.eclipse.sirius.web.spring.graphql.ws.GraphQLWebSocketHandler.handleTextMessage(GraphQLWebSocketHandler.java:260)
	at org.springframework.web.socket.handler.AbstractWebSocketHandler.handleMessage(AbstractWebSocketHandler.java:43)
	at org.springframework.web.socket.handler.WebSocketHandlerDecorator.handleMessage(WebSocketHandlerDecorator.java:75)
	at org.springframework.web.socket.handler.LoggingWebSocketHandlerDecorator.handleMessage(LoggingWebSocketHandlerDecorator.java:56)
	at org.springframework.web.socket.handler.ExceptionWebSocketHandlerDecorator.handleMessage(ExceptionWebSocketHandlerDecorator.java:58)
	at org.springframework.web.socket.adapter.standard.StandardWebSocketHandlerAdapter.handleTextMessage(StandardWebSocketHandlerAdapter.java:114)
	at org.springframework.web.socket.adapter.standard.StandardWebSocketHandlerAdapter.access$000(StandardWebSocketHandlerAdapter.java:43)
	at org.springframework.web.socket.adapter.standard.StandardWebSocketHandlerAdapter$3.onMessage(StandardWebSocketHandlerAdapter.java:85)
	at org.springframework.web.socket.adapter.standard.StandardWebSocketHandlerAdapter$3.onMessage(StandardWebSocketHandlerAdapter.java:82)
	at org.apache.tomcat.websocket.WsFrameBase.sendMessageText(WsFrameBase.java:415)
	at org.apache.tomcat.websocket.server.WsFrameServer.sendMessageText(WsFrameServer.java:129)
	at org.apache.tomcat.websocket.WsFrameBase.processDataText(WsFrameBase.java:515)
	at org.apache.tomcat.websocket.WsFrameBase.processData(WsFrameBase.java:301)
	at org.apache.tomcat.websocket.WsFrameBase.processInputBuffer(WsFrameBase.java:133)
	at org.apache.tomcat.websocket.server.WsFrameServer.onDataAvailable(WsFrameServer.java:85)
	at org.apache.tomcat.websocket.server.WsFrameServer.doOnDataAvailable(WsFrameServer.java:183)
	at org.apache.tomcat.websocket.server.WsFrameServer.notifyDataAvailable(WsFrameServer.java:162)
	at org.apache.tomcat.websocket.server.WsHttpUpgradeHandler.upgradeDispatch(WsHttpUpgradeHandler.java:156)
	at org.apache.coyote.http11.upgrade.UpgradeProcessorInternal.dispatch(UpgradeProcessorInternal.java:60)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:59)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:893)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1723)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: Not a JSON Array: {"id":"cbcb071a-0d7b-47b2-9486-7040c0cdf354","eClass":"sampledomain:Child"}
	at com.google.gson.JsonElement.getAsJsonArray(JsonElement.java:107)
	at org.eclipse.sirius.emfjson.utils.GsonEObjectDeserializer.deserializeMultipleContainmentEReference(GsonEObjectDeserializer.java:731)
	at org.eclipse.sirius.emfjson.utils.GsonEObjectDeserializer.deserializeEReference(GsonEObjectDeserializer.java:447)
	at org.eclipse.sirius.emfjson.utils.GsonEObjectDeserializer.deserializeData(GsonEObjectDeserializer.java:428)
	at org.eclipse.sirius.emfjson.utils.GsonEObjectDeserializer.deserializeEClassifier(GsonEObjectDeserializer.java:404)
	at org.eclipse.sirius.emfjson.utils.GsonEObjectDeserializer.loadObject(GsonEObjectDeserializer.java:367)
	at org.eclipse.sirius.emfjson.utils.GsonEObjectDeserializer.deserializeContent(GsonEObjectDeserializer.java:270)
	at org.eclipse.sirius.emfjson.utils.GsonEObjectDeserializer.deserialize(GsonEObjectDeserializer.java:253)
	at org.eclipse.sirius.emfjson.utils.GsonEObjectDeserializer.deserialize(GsonEObjectDeserializer.java:1)
	at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:69)
	at com.google.gson.Gson.fromJson(Gson.java:932)
	... 44 more
```

### Analysis

When loading the previously saved document, Sirius EMF JSON sees that the `child` feature in the current definition of the metamodel is multi-valued, so it expects to read the values in a JSON array. But at the time the document was saved the feature was single-valued so it was serialized as a JSON object.

### Expected Behavior

* When reading a multi-valued feature, Sirius EMF JSON should accept (maybe only if a particular option is set) single values.
* Conversely, when reading a single-valued feature, Sirius EMF JSON should accept multiple values (arrays), and only consider the first one (or consider the feature not to be set if the array is empty).

**Comment by pcdavid** [comment]
A first patch has been merged which makes EMF JSON's attributes handling more flexible, but before this issue can be closed I think we should:
- [ ] add the same kind of support to references handling.
- [ ] make the new behavior opt-out via an option. It really changes the behavior and in some cases can lead to data loss (when loading previously multi-valued feature with a metamodel where they are now single-valued), and in some use cases clients might prefer a clear error/crash than a silent loss of data.
- [ ] bump the Sirius EMF JSON version, build/release it, and consume the new version in Sirius Components
- [ ] add some tests, at least in Sirius Components to validate the scenario described above

**Comment by pcdavid** [comment]
See https://github.com/eclipse-sirius/sirius-emf-json/pull/4 for reference support.

**Comment by pcdavid** [comment]
See https://github.com/eclipse-sirius/sirius-emf-json/pull/5 for the bump

**Comment by sbegaudeau** [comment]
Merged with the switch to EMF Json 2.3.1
